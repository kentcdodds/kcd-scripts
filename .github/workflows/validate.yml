name: validate
on: [push, pull_request]
jobs:
  main:
    strategy:
      matrix:
        os: [ubuntu-latest, windows-latest]
        node: [10.14, 12, 14, 15]
    runs-on: ${{ matrix.os }}
    steps:
      - name: ⬇️ Checkout repo
        uses: actions/checkout@v2

      - name: ⎔ Setup node
        uses: actions/setup-node@v1
        with:
          node-version: ${{ matrix.node }}

      - name: 📥 Download deps
        uses: bahmutov/npm-install@v1
        with:
          useLockFile: false

      - name: ▶️ Run validate script
        run: npm run validate

      - name: ⬆️ Upload coverage report
        uses: codecov/codecov-action@v1

  release:
    needs: main
    runs-on: ubuntu-latest
    if:
      ${{ github.repository == 'kentcdodds/kcd-scripts' &&
      contains('refs/heads/master,refs/heads/beta,refs/heads/next,refs/heads/alpha',
      github.ref) && github.event_name == 'push' }}
    steps:
      - name: ⬇️ Checkout repo
        uses: actions/checkout@v2

      - name: ⎔ Setup node
        uses: actions/setup-node@v1
        with:
          node-version: 14

      - name: 📥 Download deps
        uses: bahmutov/npm-install@v1
        with:
          useLockFile: false

      - name: 🏗 Run build script
        run: npm run build

      - name: 🚀 Release
        uses: cycjimmy/semantic-release-action@v2
        with:
          semantic_version: 17
          branches: |
            [
              '+([0-9])?(.{+([0-9]),x}).x',
              'main',
              'next',
              'next-major',
              {name: 'beta', prerelease: true},
              {name: 'alpha', prerelease: true}
            ]
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
