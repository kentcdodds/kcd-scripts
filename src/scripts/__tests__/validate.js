import cases from 'jest-in-case'
import {unquoteSerializer} from './helpers/serializers'

expect.addSnapshotSerializer(unquoteSerializer)

cases(
  'validate',
  ({setup = () => () => {}}) => {
    // beforeEach
    const {sync: crossSpawnSyncMock} = require('cross-spawn')
    const originalExit = process.exit
    process.exit = jest.fn()
    process.env['SCRIPTS_PRE-COMMIT'] = 'false'
    const teardown = setup()

    try {
      // tests
      crossSpawnSyncMock.mockClear()
      require('../validate')
      expect(crossSpawnSyncMock).toHaveBeenCalledTimes(1)
      const [firstCall] = crossSpawnSyncMock.mock.calls
      const [script, calledArgs] = firstCall
      expect([script, ...calledArgs].join(' ')).toMatchSnapshot()
    } catch (error) {
      throw error
    } finally {
      teardown()
    }

    // afterEach
    process.exit = originalExit
    jest.resetModules()
  },
  {
    'calls concurrently with all scripts': {
      setup: withDefaultSetup(setupWithScripts()),
    },
    [`does not include "lint" if it doesn't have that script`]: {
      setup: withDefaultSetup(setupWithScripts(['test', 'build', 'flow'])),
    },
    [`does not include "test" if it doesn't have that script`]: {
      setup: withDefaultSetup(setupWithScripts(['lint', 'build', 'flow'])),
    },
    [`does not include "build" if it doesn't have that script`]: {
      setup: withDefaultSetup(setupWithScripts(['test', 'lint', 'flow'])),
    },
    [`does not include "flow" if it doesn't have that script`]: {
      setup: withDefaultSetup(setupWithScripts(['test', 'build', 'lint'])),
    },
    'allows you to specify your own npm scripts': {
      setup: setupWithArgs(['specialbuild,specialtest,speciallint']),
    },
    [`doesn't use test or lint if it's in pre-commit`]: {
      setup: withDefaultSetup(() => {
        const previousVal = process.env['SCRIPTS_PRE-COMMIT']
        process.env['SCRIPTS_PRE-COMMIT'] = 'true'
        return function teardown() {
          process.env['SCRIPTS_PRE-COMMIT'] = previousVal
        }
      }),
    },
  },
)

function setupWithScripts(scripts = ['test', 'lint', 'build', 'flow']) {
  return function setup() {
    const utils = require('../../utils')
    const originalIfScript = utils.ifScript
    utils.ifScript = (script, t, f) => (scripts.includes(script) ? t : f)
    return function teardown() {
      utils.ifScript = originalIfScript
    }
  }
}

function setupWithArgs(args = []) {
  return function setup() {
    const utils = require('../../utils')
    const originalResolveBin = utils.resolveBin
    utils.resolveBin = (modName, {executable = modName} = {}) => executable
    const originalArgv = process.argv
    process.argv = ['node', '../format', ...args]
    return function teardown() {
      process.argv = originalArgv
      utils.resolveBin = originalResolveBin
    }
  }
}

function withDefaultSetup(setupFn) {
  return function defaultSetup() {
    const utils = require('../../utils')
    utils.resolveBin = (modName, {executable = modName} = {}) => executable
    const argsTeardown = setupWithArgs()()
    const teardownScripts = setupWithScripts()()
    const teardownFn = setupFn()
    return function defaultTeardown() {
      argsTeardown()
      teardownFn()
      teardownScripts()
    }
  }
}
