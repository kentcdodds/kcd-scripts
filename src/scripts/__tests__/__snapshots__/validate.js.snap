// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate allows you to specify your own npm scripts 1`] = `concurrently --kill-others-on-fail --prefix [{name}] --names specialbuild,specialtest,speciallint --prefix-colors bgBlue.bold.reset,bgGreen.bold.reset,bgMagenta.bold.reset "npm run specialbuild --silent" "npm run specialtest --silent" "npm run speciallint --silent"`;

exports[`validate calls concurrently with all scripts 1`] = `concurrently --kill-others-on-fail --prefix [{name}] --names build,lint,test,flow --prefix-colors bgBlue.bold.reset,bgGreen.bold.reset,bgMagenta.bold.reset,bgCyan.bold.reset "npm run build --silent" "npm run lint --silent" "npm run test --silent -- --coverage" "npm run flow --silent"`;

exports[`validate does not include "build" if it doesn't have that script 1`] = `concurrently --kill-others-on-fail --prefix [{name}] --names lint,test,flow --prefix-colors bgBlue.bold.reset,bgGreen.bold.reset,bgMagenta.bold.reset "npm run lint --silent" "npm run test --silent -- --coverage" "npm run flow --silent"`;

exports[`validate does not include "flow" if it doesn't have that script 1`] = `concurrently --kill-others-on-fail --prefix [{name}] --names build,lint,test --prefix-colors bgBlue.bold.reset,bgGreen.bold.reset,bgMagenta.bold.reset "npm run build --silent" "npm run lint --silent" "npm run test --silent -- --coverage"`;

exports[`validate does not include "lint" if it doesn't have that script 1`] = `concurrently --kill-others-on-fail --prefix [{name}] --names build,test,flow --prefix-colors bgBlue.bold.reset,bgGreen.bold.reset,bgMagenta.bold.reset "npm run build --silent" "npm run test --silent -- --coverage" "npm run flow --silent"`;

exports[`validate does not include "test" if it doesn't have that script 1`] = `concurrently --kill-others-on-fail --prefix [{name}] --names build,lint,flow --prefix-colors bgBlue.bold.reset,bgGreen.bold.reset,bgMagenta.bold.reset "npm run build --silent" "npm run lint --silent" "npm run flow --silent"`;

exports[`validate doesn't use test or lint if it's in pre-commit 1`] = `concurrently --kill-others-on-fail --prefix [{name}] --names build,flow --prefix-colors bgBlue.bold.reset,bgGreen.bold.reset "npm run build --silent" "npm run flow --silent"`;
